Gimbal Controller Plugin[WIP]

addy1997 opened the pull request:
## Summary
Note: This PR is followed from (#840)
A model plugin to control the `gimbal` added. This plugin has been adapted from the `gimbal controller` from this [repo](https://github.com/utiasSTARS/PhoenixDrone-sitl_gazebo/blob/master/include/gazebo_gimbal_controller_plugin.hh) and the `gimbal_small_2d` plugin from the `Gazebo (classic)`.

## Test it

* In order to test the plugin, please use the gimbal model found [here](https://app.ignitionrobotics.org/OpenRobotics/fuel/models/Gimbal%20Small%202D).

* Broadly, the idea is to attach a gimbal to the base link of a drone and control it. I am still working on it and will add the files for the model on the fuel shortly.

## Checklist
- [x] Signed all commits for DCO
- [x] Added tests
- [x] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

addy1997 commented:
@chapulina, any updates on this? Is this implemented correctly?

addy1997 commented:
> ## Summary
> 
> Note: This PR is followed from (#840)
> A model plugin to control the `gimbal` added. This plugin has been adapted from the `gimbal controller` from this [repo](https://github.com/utiasSTARS/PhoenixDrone-sitl_gazebo/blob/master/include/gazebo_gimbal_controller_plugin.hh) and the `gimbal_small_2d` plugin from the `Gazebo (classic)`.
> ## Test it
> 
>     * In order to test the plugin, please use the gimbal model found [here](https://app.ignitionrobotics.org/OpenRobotics/fuel/models/Gimbal%20Small%202D).
> 
>     * Broadly, the idea is to attach a gimbal to the base link of a drone and control it. I am still working on it and will add the files for the model on the fuel shortly.
> 
> 
> ## Checklist
> 
>     * [x]  Signed all commits for DCO
> 
>     * [ ]  Added tests
> 
>     * [x]  Added example and/or tutorial
> 
>     * [x]  Updated documentation (as needed)
> 
>     * [ ]  Updated migration guide (as needed)
> 
>     * [ ]  `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
> 
>     * [ ]  All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
> 
>     * [ ]  While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

@chapulina, is this of some use?

chapulina commented:
> @chapulina, is this of some use?

I haven't had time to review this PR yet, but in general, a gimbal controller would be a great addition to Ignition Gazebo.

My preference would be to add it to `src/systems` instead of examples - the reasoning is that this would be something that we install and downstream users can easily insert into their simulations.

The bar for getting a plugin into the source folder, as opposed to examples, is a bit higher though. The plugin must have a test and example world demonstrating the functionality.

Do you have an example of this plugin working? Maybe a gif of the gimbal model in movement?

addy1997 commented:
> > @chapulina, is this of some use?
> 
> I haven't had time to review this PR yet, but in general, a gimbal controller would be a great addition to Ignition Gazebo.
> 
> My preference would be to add it to `src/systems` instead of examples - the reasoning is that this would be something that we install and downstream users can easily insert into their simulations.
> 
> The bar for getting a plugin into the source folder, as opposed to examples, is a bit higher though. The plugin must have a test and example world demonstrating the functionality.
> 
> Do you have an example of this plugin working? Maybe a gif of the gimbal model in movement?

There's an [iris](https://app.ignitionrobotics.org/OpenRobotics/fuel/models/Iris%20with%20Standoffs%20and%20Camera%20LiftDrag%20ArduCopter%20Plugins) which can be used to test this plugin but it requires the arducopter plugin which hasn't been implemented yet. @Chapulina, if my PR has flustered you, my apologies. Maybe I can draft this.

ahcorde commented on examples/plugin/gimbal_controller/GimbalControllerPlugin.cc:
```diff
@@ -0,0 +1,450 @@
+  public: void Init(const EntityComponentManager &_ecm);
+
+  /// \brief callback for imu sensor
+  public: void imuCb(const ignition::msgs::IMU &_msg);
```
doc argument

ahcorde commented on examples/plugin/gimbal_controller/GimbalControllerPlugin.cc:
```diff
@@ -0,0 +1,450 @@
+  public: std::chrono::steady_clock::duration lastControllerUpdateTime{0};
+
+  /// \brief PID
+  public: ignition::math::PID pitchPid;
+  public: ignition::math::PID yawPid;
+  public: ignition::math::PID rollPid;
```
style

ahcorde commented on examples/plugin/gimbal_controller/GimbalControllerPlugin.cc:
```diff
@@ -0,0 +1,450 @@
+
+	if(!this->dataPtr->yawJoint)
+	{
+		ignerr << "GimbalControllerPlugin::Configure ERROR! Can't get yaw joint. " <<yawJointName <<"' " <<std::endl;
```
No more than 100 characters

ahcorde commented on examples/plugin/gimbal_controller/GimbalControllerPlugin.cc:
```diff
@@ -0,0 +1,450 @@
+void GimbalControllerPluginPrivate::Init(const EntityComponentManager &_ecm)
+{    
+  // receive pitch command via ignition transport 
+	std::string pitchTopic = std::string("~/") +  this->model.Name(_ecm) + "/gimbal_pitch_cmd";
```
indentation

ahcorde commented on examples/plugin/gimbal_controller/GimbalControllerPlugin.cc:
```diff
@@ -0,0 +1,450 @@
+void GimbalControllerPluginPrivate::Init(const EntityComponentManager &_ecm)
+{    
+  // receive pitch command via ignition transport 
+	std::string pitchTopic = std::string("~/") +  this->model.Name(_ecm) + "/gimbal_pitch_cmd";
```
this symbol `~/` looks like a ROS like topic, please fix this

addy1997 replied:
you want me to remove this?

ahcorde commented on examples/plugin/gimbal_controller/GimbalControllerPlugin.cc:
```diff
@@ -0,0 +1,450 @@
+////////////////////////////////////////////////////////////////////////////
+void GimbalControllerPluginPrivate::OnPitchStringMsg(const msgs::StringMsg &_msg)
+{
+	this->pitchCommand = atof(_msg.data().c_str());
```
indentation

ahcorde commented on examples/plugin/gimbal_controller/GimbalControllerPlugin.cc:
```diff
@@ -0,0 +1,450 @@
+////////////////////////////////////////////////////////////////////////////
+void GimbalControllerPluginPrivate::OnRollStringMsg(const msgs::StringMsg &_msg)
+{
+	this->rollCommand = atof(_msg.data().c_str());
```
indentation

ahcorde commented on examples/plugin/gimbal_controller/GimbalControllerPlugin.cc:
```diff
@@ -0,0 +1,450 @@
+////////////////////////////////////////////////////////////////////////////
+void GimbalControllerPluginPrivate::OnYawStringMsg(const msgs::StringMsg &_msg)
+{
+	this->yawCommand = atof(_msg.data().c_str());
```
indentation

ahcorde commented on examples/plugin/gimbal_controller/GimbalControllerPlugin.cc:
```diff
@@ -0,0 +1,450 @@
+  double r11, double r12, double r21, double r31, double r32)
+{
+  return ignition::math::Vector3d(
+    atan2( r31, r32 ),
+    asin ( r21 ),
+    atan2( r11, r12 ));
```
```suggestion
    atan2(r31, r32),
    asin(r21),
    atan2(r11, r12));
```

ahcorde commented on examples/plugin/gimbal_controller/GimbalControllerPlugin.cc:
```diff
@@ -0,0 +1,450 @@
+////////////////////////////////////////////////////////////////////////////
+void GimbalControllerPlugin::PreUpdate(const ignition::gazebo::UpdateInfo &_info,
+    ignition::gazebo::EntityComponentManager &_ecm)
+{  if(!this->dataPtr->imuInitialized)
```
syle and indentation

addy1997 commented on examples/plugin/gimbal_controller/GimbalControllerPlugin.cc:
```diff
@@ -0,0 +1,450 @@
+class ignition::gazebo::systems::GimbalControllerPluginPrivate
+{    
+  /// \brief function for Pitch joint message
+  public: void OnPitchStringMsg(const msgs::StringMsg &_msg);
```
?

ahcorde replied:
I was refering here to the documentation of the method

```
    /// \brief
    /// \param[in] <var_name> description
    /// \param[in] 
    /// \return explain the return value if exists
```

ahcorde reviewed (changes requested):
line should not be longer than 80 characters, indentation should be the same in the hole code.

You should add:
 - some tests to check than the implementation is working (you can check other plugins, for example this [test](https://github.com/ignitionrobotics/ign-gazebo/blob/ign-gazebo5/test/integration/joint_position_controller_system.cc))
 - Add a example world [here](https://github.com/ignitionrobotics/ign-gazebo/tree/ign-gazebo5/examples/worlds)

addy1997 commented on examples/plugin/gimbal_controller/GimbalControllerPlugin.cc:
```diff
@@ -0,0 +1,450 @@
+          ss << yawJointPos;
+          m.set_data(ss.str());
+          this->dataPtr->yawPub.Publish(m);
+       }
+  }
```
done sir

addy1997 commented on examples/plugin/gimbal_controller/GimbalControllerPlugin.cc:
```diff
@@ -0,0 +1,450 @@
+          rjfcComp = _ecm.Component<components::JointForceCmd>(_ecm.CreateComponent(this->dataPtr->rollJoint, components::JointForceCmd({0})));
+       }
+
+       static int i =1000;
+       if (++i>100)
```
done

addy1997 commented:
> line should not be longer than 80 characters, indentation should be the same in the hole code.
> 
> You should add:
> 
>     * some tests to check than the implementation is working (you can check other plugins, for example this [test](https://github.com/ignitionrobotics/ign-gazebo/blob/ign-gazebo5/test/integration/joint_position_controller_system.cc))
> 
>     * Add a example world [here](https://github.com/ignitionrobotics/ign-gazebo/tree/ign-gazebo5/examples/worlds)

@ahcorde, Is there a fixed template for writing a test for the model plugin. Also, the world example for testing the plugin can be found [here](https://app.ignitionrobotics.org/OpenRobotics/fuel/models/Iris%20with%20Standoffs%20and%20Camera%20LiftDrag%20ArduCopter%20Plugin)

ahcorde commented:
> Is there a fixed template for writing a test for the model plugin ?

As far as I know there is no template, but you based your test in one of the tests in this [folder](ttps://github.com/ignitionrobotics/ign-gazebo/blob/ign-gazebo5/test/integration/)

> the world example for testing the plugin can be found here

The example that you mentioned includes this plugin ? I would recomend you to create a new file in the example world folder with a working example.
By the way you will need to create one sdf file for your test.

ahcorde commented:
You don't need to add a super complex sdf, you can add a simple model to test your new plugin

addy1997 commented:
> The example that you mentioned includes this plugin ? 

Yes, it does. Actually, Gimbal is attached to the iris drone.

ahcorde reviewed (changes requested):
> Yes, it does. Actually, Gimbal is attached to the iris drone.

But I mean is this **new** plugin included in the model ? You should modify the URDF, right ? 

Anyhow, for this PR I expect:

 - A new test for the gimbal which includes: 
    - a new file in the integration folder with the test
    - a sdf file with a simple gimbal. You don't need to add meshes just a simple mechanism. This file should be loaded by the test
 - A example world with a gimbal with some instrucctions at the top of file with some commands to test you plugin.

addy1997 commented:
> > Yes, it does. Actually, Gimbal is attached to the iris drone.
> 
> But I mean is this **new** plugin included in the model ? You should modify the URDF, right ?
> 
> Anyhow, for this PR I expect:
> 
>     * A new test for the gimbal which includes:
>       
>       * a new file in the integration folder with the test
>       * a sdf file with a simple gimbal. You don't need to add meshes just a simple mechanism. This file should be loaded by the test
> 
>     * A example world with a gimbal with some instrucctions at the top of file with some commands to test you plugin.

@ahcorde, I am on it, sir.

addy1997 commented:
> > Yes, it does. Actually, Gimbal is attached to the iris drone.
> 
> But I mean is this **new** plugin included in the model ? You should modify the URDF, right ?
> 
> Anyhow, for this PR I expect:
> 
>     * A new test for the gimbal which includes:
>       
>       * a new file in the integration folder with the test
>       * a sdf file with a simple gimbal. You don't need to add meshes just a simple mechanism. This file should be loaded by the test
> 
>     * A example world with a gimbal with some instrucctions at the top of file with some commands to test you plugin.

Update: The world example for gimbal involves a drone which will need me to add load libArducopterPlugin.so file as well. Could you please check whether this [PR](https://github.com/ignitionrobotics/ign-gazebo/pull/911) (Arducopter Plugin) is implemented properly?

ahcorde commented on examples/plugin/gimbal_controller/GimbalControllerPlugin.hh:
```diff
@@ -0,0 +1,63 @@
+       public: GimbalControllerPlugin();
+
+       /// \brief Destructor.
+       public: ~GimbalControllerPlugin();
```
where is the destructor defined ?

ahcorde commented on examples/plugin/gimbal_controller/GimbalControllerPlugin.cc:
```diff
@@ -0,0 +1,461 @@
+   }
+
+   // for pitch joint
+   std::string pitchJointName = "cgo3_camera_joint";
```
these name looks gimbal specific, you should  use something generic

ahcorde reviewed (changes requested):
Hi @addy1997,

thank you for putting some time in the tests. I want to clarify some points here:

 - The plugin that you are adding is a generic gimbal controller, which means that it could be use by arducopter or some other application. I could undestand that It could be easy to test with a third party library/executable but this plugin should be completely independent. 

 - Ardupilot is a third party software not related with ignition. We don't want to add this dependency to our stack. Because If we want to test the gimbal we will need to install the hole ardupilot stack which is something that will add a big overhead to our infraestructure.

 - Ardupilot has a initial port in an external repository [here] (https://github.com/gerkey/ardupilot_gazebo/tree/ignition).

**Action points :** 

 - Close the ardupilot PR, maybe these code should live in a different repository.
 - Move the plugin from `examples` to `src/systems`. 
 - make it work with the test file that I have added. Feel free to modify the gimbal.sdf as much as you want. I mean name of the joint, links, etc, but please don not add a drone.

I have create this branch https://github.com/ignitionrobotics/ign-gazebo/tree/gimbal-cont with some changes. Please cherry-pick this commit and continue from this point

The SDF file contain this simple gimbal. Please if you have some doubts ask them, I will be happy to answer

![gimbal](https://user-images.githubusercontent.com/1933907/125271392-fe445000-e30a-11eb-84d7-01e8ea3c3776.gif)

addy1997 commented:
@ahcorde, few updates

1. #911 closed
2. I have successfully merged your commit(https://github.com/ignitionrobotics/ign-gazebo/pull/845/commits/1893001c72ad81459523c84e5ff5a966611adb95) into this branch,
3. Have made some changes as you suggested above. A few changes are yet to be made. 
4. Tested the **gimbal.sdf** file you added. It worked fine on **ign gazebo edifice**(gif below). Apologies for glitches.

![gimbal](https://user-images.githubusercontent.com/29406906/125311735-438b7080-e351-11eb-909c-046246ce2037.gif)

ahcorde reviewed (changes requested):
There are a bunch of [linters errors](https://github.com/ignitionrobotics/ign-gazebo/pull/845/checks?check_run_id=3047282009)

addy1997 commented on examples/plugin/gimbal_controller/GimbalControllerPlugin.cc:
```diff
@@ -0,0 +1,461 @@
+      // for pitch 
+      auto pitchJointPos = _ecm.Component<components::JointPosition>(this->dataPtr->pitchJoint);
+      double pitchError = pitchJointPos->Data().at(0) - this->dataPtr->pitchCommand;
+      double pitchForce = this->dataPtr->pitchPid.Update(pitchError, dt);
```
pitchJointPos is used (in line 447)

addy1997 commented on test/worlds/gimbal.sdf:
```diff
@@ -0,0 +1,379 @@
+
+      <plugin filename="ignition-gazebo-gimbalcontrollerplugin-system"
+        name="ignition::gazebo::systems::GimbalControllerPlugin">
+        <joint_yaw></joint_yaw>
```
@ahcorde, here's the update

1. Test files are added (along with the test world)
2. IMU is added to both the SDF files.
3. All linter errors mentioned [here](https://github.com/ignitionrobotics/ign-gazebo/pull/845/checks?check_run_id=3047282009) have been fixed.

**Issue**

While simulating the gimbal.sdf file after adding IMU, I have faced these issues (given below). 
```
[Wrn] [GimbalControllerPlugin.cc:227] joint_roll [rev_joint_2] does not exist?

 [Err] [GimbalControllerPlugin.cc:233] GimbalControllerPlugin::Configure ERROR! Can't get roll joint.rev_joint_2'
```
https://user-images.githubusercontent.com/29406906/125634892-b2bd0905-3b89-4be2-990e-420d5ae9cd73.mov

The plugin loads successfully but it doesn't seem to work as anticipated. Please could you suggest something?

Thanks

addy1997 commented:
@ahcorde, here's the update

1. Test files are added (along with the test world)
2. IMU is added to both the SDF files.
3. All linter errors mentioned [here](https://github.com/ignitionrobotics/ign-gazebo/pull/845/checks?check_run_id=3047282009) have been fixed.

**Issue**

While simulating the gimbal.sdf file after adding IMU, I have faced these issues (given below). 
```
[Wrn] [GimbalControllerPlugin.cc:227] joint_roll [rev_joint_2] does not exist?

 [Err] [GimbalControllerPlugin.cc:233] GimbalControllerPlugin::Configure ERROR! Can't get roll joint.rev_joint_2'
```
https://user-images.githubusercontent.com/29406906/125634892-b2bd0905-3b89-4be2-990e-420d5ae9cd73.mov

The plugin loads successfully but it doesn't seem to work as anticipated. Please could you suggest something?
```
[Dbg] [SimulationRunner.cc:851] Loaded system [ignition::gazebo::systems::GimbalControllerPlugin] for entity [8]

```

Thanks

ahcorde commented on src/systems/gimbal_controller/GimbalControllerPlugin.cc:
```diff
@@ -0,0 +1,490 @@
+      }
+
+      static int i = 1000;
+      if (++i > 100)
```
why are you publishing 100 times the same value ?

addy1997 replied:
Referring to the plugin [file](https://github.com/addy1997/gazebo/blob/gazebo11/plugins/GimbalSmall2dPlugin.cc) in the gazebo classic, I think this portion is for publishing the **message** contents in form of string.

ahcorde commented on src/systems/gimbal_controller/GimbalControllerPlugin.cc:
```diff
@@ -0,0 +1,490 @@
+
+      if (pjfcComp == nullptr)
+      {
+         pjfcComp = _ecm.Component<components::JointForceCmd>(
+            _ecm.CreateComponent(this->dataPtr->pitchJoint,
+               components::JointForceCmd({0})));
```
I think this should looks like

```suggestion
        *pjfcComp = ignition::gazebo::components::JointForceCmd(
          <joint_force_cmd>);
```

addy1997 commented on examples/worlds/gimbal.sdf:
```diff
@@ -0,0 +1,381 @@
+        name="ignition::gazebo::systems::GimbalControllerPlugin">
+        <joint_yaw>yaw_joint</joint_yaw>
+        <joint_roll>rev_joint_2</joint_roll>
+        <joint_pitch>fix_joint_1</joint_pitch>
```
If I am not wrong, this plugin tilts the roll joint only, right? Well, I have replaced the fix_joint_1 with rev_joint_1.

addy1997 commented:
Due to some issues with my **GitHub desktop client**, I added some unwanted commits. This PR needs to be fixed.

addy1997 commented:
I had to close #845 due to some issues from my side. I'll open another PR
with improvement. Apologies for the same.

On Wed, Jul 14, 2021 at 8:39 PM Alejandro Hernández Cordero <
***@***.***> wrote:

> ***@***.**** requested changes on this pull request.
> ------------------------------
>
> In examples/worlds/gimbal.sdf
> <https://github.com/ignitionrobotics/ign-gazebo/pull/845#discussion_r669696437>
> :
>
> > +              <cfm>0</cfm>
>
> +              <erp>0.2</erp>
>
> +            </limit>
>
> +            <suspension>
>
> +              <cfm>0</cfm>
>
> +              <erp>0.2</erp>
>
> +            </suspension>
>
> +          </ode>
>
> +        </physics>
>
> +      </joint>
>
> +
>
> +      <plugin filename="ignition-gazebo-gimbalcontrollerplugin-system"
>
> +        name="ignition::gazebo::systems::GimbalControllerPlugin">
>
> +        <joint_yaw>yaw_joint</joint_yaw>
>
> +        <joint_roll>rev_joint_2</joint_roll>
>
> +        <joint_pitch>fix_joint_1</joint_pitch>
>
>
> This looks wrong, You are trying to move a fixed joint.
> ------------------------------
>
> In src/systems/gimbal_controller/GimbalControllerPlugin.cc
> <https://github.com/ignitionrobotics/ign-gazebo/pull/845#discussion_r669705196>
> :
>
> > +      auto rollJointPos = _ecm.Component<components::JointPosition>(
>
> +         this->dataPtr->rollJoint)->Data().at{0};
>
> +
>
> +      ignition::gazebo::components::JointForceCmd* rjfcComp = nullptr;
>
> +      rjfcComp = _ecm.Component<components::JointForceCmd>(
>
> +         this->dataPtr->rollJoint);
>
> +
>
> +      if (rjfcComp == nullptr)
>
> +      {
>
> +         rjfcComp = _ecm.Component<components::JointForceCmd>(
>
> +         _ecm.CreateComponent(this->dataPtr->rollJoint,
>
> +            components::JointForceCmd({0})));
>
> +      }
>
> +
>
> +      static int i = 1000;
>
> +      if (++i > 100)
>
>
> why are you publishing 100 times the same value ?
> ------------------------------
>
> In src/systems/gimbal_controller/GimbalControllerPlugin.cc
> <https://github.com/ignitionrobotics/ign-gazebo/pull/845#discussion_r669706210>
> :
>
> > +         pjfcComp = _ecm.Component<components::JointForceCmd>(
>
> +            _ecm.CreateComponent(this->dataPtr->pitchJoint,
>
> +               components::JointForceCmd({0})));
>
>
> I think this should looks like
> ⬇️ Suggested change
>
> -         pjfcComp = _ecm.Component<components::JointForceCmd>(
>
> -            _ecm.CreateComponent(this->dataPtr->pitchJoint,
>
> -               components::JointForceCmd({0})));
>
> +        *pjfcComp = ignition::gazebo::components::JointForceCmd(
>
> +          <joint_force_cmd>);
>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ignitionrobotics/ign-gazebo/pull/845#pullrequestreview-706383314>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHALNOR74HNI7Y5E6GAP5J3TXWSCRANCNFSM456E3YKQ>
> .
>


-- 
Adwait <https://www.linkedin.com/in/adwait-naik-86586514a/>
Adwait <https://github.com/addy1997>

ahcorde commented:
@addy1997, I spend time reviewing this PR, do you mind to reopen it and continue working on it? before working in others?

addy1997 commented:
@ahcorde, I will open the PR, Sir. Apologies for this but I thought that I
am disturbing you unnecessarily.

On Mon, 19 Jul 2021, 12:51 Alejandro Hernández Cordero, <
***@***.***> wrote:

> @addy1997 <https://github.com/addy1997>, I spend time reviewing this PR,
> do you mind to reopen it and continue working on it? before working in
> others?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ignitionrobotics/ign-gazebo/pull/845#issuecomment-882307033>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHALNOVITTUEUEZMYYAD5GDTYPG6BANCNFSM456E3YKQ>
> .
>

addy1997 commented:
@ahcorde, do you want me to open a fresh PR with all new files including
the test files?

On Mon, 19 Jul 2021, 14:17 Adwait Naik, ***@***.***> wrote:

> @ahcorde, I will open the PR, Sir. Apologies for this but I thought that I
> am disturbing you unnecessarily.
>
> On Mon, 19 Jul 2021, 12:51 Alejandro Hernández Cordero, <
> ***@***.***> wrote:
>
>> @addy1997 <https://github.com/addy1997>, I spend time reviewing this PR,
>> do you mind to reopen it and continue working on it? before working in
>> others?
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/ignitionrobotics/ign-gazebo/pull/845#issuecomment-882307033>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AHALNOVITTUEUEZMYYAD5GDTYPG6BANCNFSM456E3YKQ>
>> .
>>
>

ahcorde commented:
You don't need to open a new PR, you can rewrite the history of this commit

addy1997 commented:
> You don't need to open a new PR, you can rewrite the history of this commit

@ahcorde, I have added all the files as per your previous suggestions. 

- Two world files (one in example and one for test) have been added
- The test file has been added
- I have fixed most of the errors (including the linters). 

Should you find any files missing, please let me know.

Thanks

ahcorde reviewed (changes requested):
You have conflicts with master and code is not compiling in any platform, For example Ubuntu: https://build.osrfoundation.org/job/ignition_gazebo-ci-pr_any-ubuntu_auto-amd64/6591/consoleFull#-18463618236ea37b8d-a6d4-40ae-8431-d90c018842af

Can you provide with next update a small gif showing how the gimbal is working. You can use the tool  `peek` on Ubuntu.

addy1997 commented:
> You have conflicts with master and code is not compiling in any platform, For example Ubuntu: https://build.osrfoundation.org/job/ignition_gazebo-ci-pr_any-ubuntu_auto-amd64/6591/consoleFull#-18463618236ea37b8d-a6d4-40ae-8431-d90c018842af
> 
> Can you provide with next update a small gif showing how the gimbal is working. You can use the tool `peek` on Ubuntu.

@ahcorde, Please find the GIF below. 

![gimbal](https://user-images.githubusercontent.com/29406906/126543573-389f4de5-5c1c-40b9-92b8-2dc9aa2fa611.gif)

The plugin loads properly

![Screenshot 2021-07-22 at 00 23 30](https://user-images.githubusercontent.com/29406906/126543860-6d124e34-f54f-442a-bd03-b90d132eb3bc.png)

I am getting this error
![Screenshot 2021-07-22 at 00 23 44](https://user-images.githubusercontent.com/29406906/126543886-19ae5a62-b11f-4f3e-9dfa-d292517d9fac.png)

ahcorde reviewed (changes requested):
The gif that you shared it shows the gimbal, but it's not working, it's just moving because of the gravety. Please review the logic and I would love to see a gif of the gimbal and a terminal seding a command, then the gimbal should move to the commanded position. 

Regarding the build error, please check the CI, Ubuntu, MACOS and Windows are not building
 - Ubuntu https://build.osrfoundation.org/job/ignition_gazebo-ci-pr_any-ubuntu_auto-amd64/6591/
 - MACOS https://build.osrfoundation.org/job/ignition_gazebo-ci-pr_any-homebrew-amd64/6090/
 - Windows https://build.osrfoundation.org/job/ign_gazebo-pr-win/2778/

addy1997 commented:
@ahcorde, I did some fixes but while loading the plugin, it gives me this error

![Screenshot 2021-07-24 at 14 55 37](https://user-images.githubusercontent.com/29406906/126864212-6197d61f-c810-48b4-9566-2de63e37c357.png)

Note: I am using this command to copy plugin file

`sudo cp libignition-gazebo-gimbalcontrollerplugin-system.so /usr/lib/x86_64-linux-gnu/ign-gazebo-5/plugins`

Any idea on how to fix this issue?

addy1997 commented:
@ahcorde, if in case you have spare time, could you help me with this PR?

ahcorde reviewed (changes requested):
Please don't include image with error messages, because there are not searchable.

You don't need to copy the library in your system, you should create a workspace, compile your code, source it and then run/test your code

addy1997 commented:
> Please don't include image with error messages, because there are not searchable.
> 
> You don't need to copy the library in your system, you should create a workspace, compile your code, source it and then run/test your code
@ahcorde, I am working on this. Meanwhile, can you please review my [test file](https://github.com/ignitionrobotics/ign-gazebo/pull/845/files#diff-455d804122eb0b6b379811b2646676304aede490417124ccc982f0cbd878bf52) is correct?

addy1997 commented:
@ahcorde, how would you suggest proceeding on this PR?

ahcorde commented:
@addy1997 as far as I remember this is not working, you should:

 - make changes to make the plugin work. When you accomplish this goal, please share a gif with us.
 - then you should rewrite the test with the working plugin

chapulina commented:
@addy1997 , are you planning to keep working on this PR? Thanks.

addy1997 commented:
> @addy1997 , are you planning to keep working on this PR? Thanks.

@chapulina, I want to convert this PR into a draft till I work on the solution. Is it fine?

addy1997 commented:
@chapulina, I would require some help with this PR. I want someone to check the implementation for Gimabal.

chapulina commented:
> @chapulina, I would require some help with this PR. I want someone to check the implementation for Gimbal.

Hi @addy1997 , can you elaborate on what you've tried, what's working, and what isn't? This is quite a lot of code, so it will take time for us to review it. If you can provide us as much information as possible, that will facilitate the review.

addy1997 commented:
> > @chapulina, I would require some help with this PR. I want someone to check the implementation for Gimbal.
> 
> Hi @addy1997 , can you elaborate on what you've tried, what's working, and what isn't? This is quite a lot of code, so it will take time for us to review it. If you can provide us as much information as possible, that will facilitate the review.


Hi @chapulina, 

Sorry for the late reply. 

The following files need to be reviewed. If you are super busy, please review them at your convenience. Also my apologies for making this PR super confusing. 

1. GimbalControllerPlugin.hh #header file
2. GimbalControllerPlugin.cc #MAIN file
3. gimbal_controller_plugin.hh (file for test/integration) (as @ahcorde told me to do)
4. Gimbal.sdf(works fine, review not required)

chapulina commented on src/systems/physics/Physics.cc:
```diff
@@ -302,16 +302,20 @@ class ignition::gazebo::systems::PhysicsPrivate
+  /// \brief Feature list to handle contacts information.
+  public: struct ContactFeatureList : ignition::physics::FeatureList<
+            CollisionFeatureList,
+            ignition::physics::GetContactsFromLastStepFeature>{};
```
Are the changes to this system needed? Can they be reverted?

chapulina commented on src/systems/gimbal_controller/GimbalControllerPlugin.hh:
```diff
@@ -0,0 +1,82 @@
+    class GimbalControllerPluginPrivate;
+
+    /// \brief Plugin to control a 3-axis gimbal.
+    /// See the example usage below:
```
Can you also document the topics that are provided by this system?

Also please explain its relation to the IMU sensor and how they're expected to work together.

chapulina commented on src/systems/gimbal_controller/GimbalControllerPlugin.cc:
```diff
@@ -0,0 +1,484 @@
+  // receive yaw command via gz transport
+  std::string yawTopic = this->model.Name(_ecm) + "/gimbal_yaw_cmd";
+  this->node.Subscribe(yawTopic, 
+    &GimbalControllerPluginPrivate::OnYawStringMsg);
```
Instead of converting string to doubles, we should let users publish the doubles directly. How about offering a single topic that takes an `ignition::msgs::Vector3d` message?

chapulina commented on src/systems/gimbal_controller/GimbalControllerPlugin.cc:
```diff
@@ -0,0 +1,484 @@
+
+  // publish pitch status via gz transport
+  pitchTopic = this->model.Name(_ecm) + "/gimbal_pitch_status";
+  this->pitchPub= this->node.Advertise<ignition::msgs::StringMsg>(pitchTopic);
```
Same here about using an expressive type instead of a string

chapulina commented on src/systems/gimbal_controller/GimbalControllerPlugin.cc:
```diff
@@ -0,0 +1,484 @@
+
+//////////////////////////////////////////////////
+ignition::math::Vector3d GimbalControllerPluginPrivate::QtoZXY
+(const ignition::math::Quaterniond &_q)
```
This function isn't used, can it be removed?

Please go over all functions and variables and remove the ones which aren't necessary.

chapulina commented on examples/worlds/gimbal.sdf:
```diff
@@ -0,0 +1,390 @@
+<?xml version="1.0" ?>
+<!--
+  Ignition gazebo demo using IMU sensor for Gimbal
```
Can you add some lines explaining the user what commands they can try, and what output they should expect to see? See how we do it in other examples, for example:

https://github.com/ignitionrobotics/ign-gazebo/blob/17f63c497e7fb81dca5f87708ba8b2128db67b31/examples/worlds/ackermann_steering.sdf#L3-L11

chapulina commented on examples/worlds/gimbal.sdf:
```diff
@@ -0,0 +1,390 @@
+    <plugin
+      filename="ignition-gazebo-magnetometer-system"
+      name="ignition::gazebo::systems::Magnetometer">
+    </plugin>
```
Do we need all of these systems here just to demo the gimbal? The example should show users the minimal way of using that features, so they can take it and incorporate into their projects without bringing unnecessary things.

chapulina commented on src/systems/gimbal_controller/GimbalControllerPlugin.cc:
```diff
@@ -0,0 +1,484 @@
+
+    if(_info.simTime < this->dataPtr->lastControllerUpdateTime)
+    {    
+       ignerr << "time reset event\n";
```
Is this an error? If it is, we should be more verbose and explain to the user what will break in this case.

chapulina commented on src/systems/gimbal_controller/GimbalControllerPlugin.cc:
```diff
@@ -0,0 +1,484 @@
+       this->dataPtr->lastControllerUpdateTime = _info.simTime;
+        return;
+    }
+    else if (_info.simTime > this->dataPtr->lastControllerUpdateTime)
```
If we're returning above, this doesn't need to be inside an `else if`

chapulina commented on src/systems/gimbal_controller/GimbalControllerPlugin.cc:
```diff
@@ -0,0 +1,484 @@
+      rjfcComp = _ecm.Component<components::JointForceCmd>
+      (this->dataPtr->rollJoint);
+
+      if (rjfcComp == nullptr)
```
This is only setting the command if the component doesn't exist yet. If it exists, the command won't be set. Is that expected?

chapulina reviewed (commented):
Left a high-level review, but I haven't tried running the code.
